app/main.py

```diff
@@ -88,6 +91,37 @@ def _get_client(self):
                                              self.client_id, self.client_secret)
         return KeyVaultClient(credentials)
 
+    def _get_tenant_id(self, tenant_id_from_config):
+        if os.getenv('AUTO_DETECT_AAD_TENANT').lower() != 'true':
+            _logger.info('AAD tenant auto detection turned off. Using tenant id %s from cloud config', vault_base_url)
+            return tenant_id_from_config
+        else:
+            vault_base_url = os.getenv('VAULT_BASE_URL')
+            _logger.info('AAD tenant auto detection turned on. Detecting tenant id for %s', vault_base_url)
+            # Some key to trigger a 401
+            URL = '{}/keys/somekeyname/1?api-version=2018-02-14'.format(vault_base_url)
+            _logger.info('Sending challenge request to %s', URL)
+            response = requests.get(url = URL)
+            if response.status_code == 401:
+                # If status code == HTTP 401, then parse the WWW-Authenticate header
+                # Bearer authorization="https://login.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47", resource=".."
+                # Parse the value of authorization from the challenge
+                challenge = response.headers['WWW-authenticate'].lower()
+                _logger.info('Received challenge %s', challenge)
+                challenge_data = .replace('bearer ', '').split(',')
+                for kvp in challenge_data:
+                    keyvalue = kvp.strip().split('=')
+                    if len(keyvalue) == 2 and keyvalue[0] == 'authorization':
+                        authority = keyvalue[1].replace('"', '')
+                        tenant_id = urlparse(authority).path.replace('/', '')
+                        _logger.info('Tenant in config : %s, Tenant detected : %s. Using %s', tenant_id_from_config, tenant_id, tenant_id)
+                        return tenant_id
+                        break
+            else:
```

Comment:
can also do this without the `else`. If you you don't return a `tenant_id` treat this as an error (and then either return the tenant ID from config or abort as Paul has suggested

Otherwise, LGTM